### PR TITLE
feat(poetry): support rangeStrategy=update-lockfile

### PR DESCRIPTION
**Pull request by orf**
## Changes:

WIP: see #8671

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added unit tests, or
- [ ] No new tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by orf** [comment]
I've made the initial changes and it correctly resolves the `lockVersion`:

```json
 "deps": [
               {
                 "depName": "boto3",
                 "depType": "dependencies",
                 "currentValue": "*",
                 "managerData": {"nestedVersion": false},
                 "datasource": "pypi",
                 "lockedVersion": "1.17.5",
                 "versioning": "poetry",
                 "depIndex": 0,
                 "updates": [],
                 "warnings": [],
                 "fixedVersion": "1.17.5",
                 "sourceUrl": "https://github.com/boto/boto3"
               }
```

However it's not picking up that there are any updates (1.17.5 is not the latest version of boto3). I'm using [this test project](https://github.com/orf/test-renovate-star/) if it helps.

Any idea @rarkins?

**Comment by orf** [comment]
Ahh. Silly me, I needed to set the range strategy to `update-lockfile`. It works now, I'll clean up and add some specific tests.

File: lib/manager/poetry/extract.ts [hunk 1]
```diff
@@ -103,10 +114,29 @@ export function extractPackageFile(
     logger.debug(`${fileName} contains no poetry section`);
     return null;
   }
+
+  // handle the lockfile
+  const lockContents = await readLocalFile('poetry.lock', 'utf8');
```
Comment (rarkins) [inline lib/manager/poetry/extract.ts:119]: This assumes the lock file is always in the root of the repo. Need to get the sibling file of the package file instead

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/manager/poetry/extract.ts (hunk 1 above)
Comment (orf) [reply lib/manager/poetry/extract.ts]: Done!

File: lib/manager/poetry/__fixtures__/pyproject.10.toml [hunk 2]
```diff
@@ -1,17 +1,3 @@
-[[tool.poetry.source]]
```
Comment (orf) [inline lib/manager/poetry/__fixtures__/pyproject.10.toml:1]: This file was unused, so I've just replaced the contents

File: lib/manager/poetry/extract.spec.ts [hunk 3]
```diff
@@ -1,6 +1,9 @@
 import { readFileSync } from 'fs';
+import { fs } from '../../../test/util';
 import { extractPackageFile } from './extract';
 
+jest.mock('../../util/fs');
```
Comment (orf) [inline lib/manager/poetry/extract.spec.ts:5]: Adding this mock here causes the ` ● .updateArtifacts() › passes private credential environment vars` test to fail. I've got *no* clue about Jest or Jest mocking so I really have no idea what is up with that.

File: lib/manager/poetry/artifacts.spec.ts [hunk 4]
```diff
@@ -19,6 +19,7 @@ const pyproject10toml = readFileSync(
 
 jest.mock('fs-extra');
 jest.mock('child_process');
+jest.mock('../../util/fs');
```
Comment (viceice) [inline lib/manager/poetry/artifacts.spec.ts:22]:
```suggestion
```

already mocked by `jest.mock('fs-extra');` which is used by this.

File: lib/manager/poetry/extract.ts [hunk 5]
```diff
@@ -1,22 +1,31 @@
...
+  PoetryLock,
+  PoetryLockSection,
+  PoetrySection,
+} from './types';
 
 function extractFromSection(
   parsedFile: PoetryFile,
-  section: keyof PoetrySection
+  section: keyof PoetrySection,
+  poetryLockfile: Record<string, PoetryLockSection> = {}
```
Comment (viceice) [inline lib/manager/poetry/extract.ts:21]:
```suggestion
  poetryLockfile: Record<string, PoetryLockSection>
```

File: lib/manager/poetry/extract.spec.ts (hunk 3 above)
Comment (viceice) [inline lib/manager/poetry/extract.spec.ts:5]:
```suggestion
```

**Review by viceice — APPROVED** [review]

File: lib/manager/poetry/__fixtures__/pyproject.10.toml [hunk 6]
```diff
@@ -1,3 +1,18 @@
-[tool.poetry.dependencies]
-python = "^3.9"
-boto3 = "*"
+[[tool.poetry.source]]
+This conversation was marked as resolved by viceice
```
Comment (rarkins) [inline lib/manager/poetry/__fixtures__/pyproject.10.toml]: @viceice are you seeing what I'm seeing?

image

Comment (viceice) [reply lib/manager/poetry/__fixtures__/pyproject.10.toml]: already fixed 🙃

**Review by rarkins — APPROVED** [review]

**Comment by rarkins** [comment]
Thanks @orf 

Feel free to add yourself to the contributors list in `package.json` anytime too

**Pull request merged**